src/lib/replication/repl_dev/raft_repl_dev.cpp

```diff
@@ -1265,39 +1278,81 @@ cshared< ReplDevCPContext > RaftReplDev::get_cp_ctx(CP* cp) {
+                    get_elapsed_time_sec(rreq->created_time()));
+            // don't clean up proposer's request
+            return;
+        }
+        if (rreq->dsn() < cur_dsn) {
+            RD_LOGD("StateMachine: legacy req, rreq=[{}] , dsn = {}, next_dsn = {}, gap= {}", rreq->to_string(), rreq->dsn(), cur_dsn,
+                    cur_dsn - rreq->dsn());
+        }
         if (rreq->is_expired()) {
-            expired_keys.push_back(key);
-            RD_LOGD("rreq=[{}] is expired, cleaning up; elapsed_time_sec{};", rreq->to_string(),
+            RD_LOGD("StateMachine: rreq=[{}] is expired, elapsed_time_sec{};", rreq->to_string(),
                     get_elapsed_time_sec(rreq->created_time()));
-
-            // do garbage collection
-            // 1. free the allocated blocks
-            if (rreq->has_state(repl_req_state_t::BLK_ALLOCATED)) {
-                auto blkid = rreq->local_blkid();
-                data_service().async_free_blk(blkid).thenValue([this, blkid](auto&& err) {
-                    HS_LOG_ASSERT(!err, "freeing blkid={} upon error failed, potential to cause blk leak",
-                                  blkid.to_string());
-                    RD_LOGD("blkid={} freed successfully", blkid.to_string());
-                });
-            }
-
-            // 2. remove from the m_repl_key_req_map
-            // handle_error during fetch data response might have already removed the rreq from the this map
-            if (m_repl_key_req_map.find(rreq->rkey()) != m_repl_key_req_map.end()) {
-                m_repl_key_req_map.erase(rreq->rkey());
-            }
         }
     });
-
-    for (auto const& l : expired_keys) {
-        m_state_machine->unlink_lsn_to_req(l);
+    RD_LOGW("state_machine req map size is {};", sm_req_cnt);
+
+    for (auto removing_rreq : expired_rreqs) {
+        // do garbage collection
+        // 1. free the allocated blocks
+        RD_LOGI("Removing rreq [{}]", removing_rreq->to_string());
+        if (removing_rreq->has_state(repl_req_state_t::BLK_ALLOCATED)) {
+            auto blkid = removing_rreq->local_blkid();
+            data_service().async_free_blk(blkid).thenValue([this, blkid](auto&& err) {
+                HS_LOG_ASSERT(!err, "freeing blkid={} upon error failed, potential to cause blk leak",
+                              blkid.to_string());
+                RD_LOGD("GC rreq: Releasing blkid={} freed successfully", blkid.to_string());
+            });
+        }
+        // 2. remove from the m_repl_key_req_map
+        if (m_repl_key_req_map.find(removing_rreq->rkey()) != m_repl_key_req_map.end()) {
+            m_repl_key_req_map.erase(removing_rreq->rkey());
+        }
+        // 3. remove from state-machine
+        if (removing_rreq->has_state(repl_req_state_t::LOG_FLUSHED)) {
+            RD_LOGW("Removing rreq [{}] from state machine, it is risky")
+            m_state_machine->unlink_lsn_to_req(removing_rreq->lsn());
```

Comment:
if the log at lsn N is overwritten(rollback happens), a new rreq will be linked to lsn N.
now , if you unlink the lsn N which is hold in the stale rreq, the new rreq will be erased from statemachine, where I believe an error will happen.

Reply:
reaching here means we have <term =1, lsn =N,  DSN = K>  committed.

How can it possible we have a <term = 2, LSN = N>  in state-machine?

Reply:
Assuming we have a leader L1 , it generates a rreq <term =1, lsn =100, DSN = 110>, now it`s `m_next_dsn` is 111. 
well, in the two followers, `m_next_dsn` is 100, since they do not receive rreq with dsn 101-110.  (for example , L1 is network partitioned). so in the leader`s memory ,  rreqs(term 1) with dsn 101-110  exists.

now , leader switch to a new one L2,  it generates  rreqs from <term =2, lsn=91, DSN = 101> to  <term =2, lsn=100, DSN = 110>  and they are committed.

if L1 reconnects to the network, it will received rreqs from <term =2, lsn=91, DSN = 101> to  <term =2, lsn=100, DSN = 110>,  it will create new rreq with dsn 101-110 and lsn 91-100 in memory , but please note that the `m_next_dsn` in L1 is 111 . now, if the stale(term 1 and expired) rreqs with dsn 101-110(they are less than 111) is gc, according to the code here, they will unlink lsn 91-100 in memory, which is now linked to the new rreq(term 2) with dsn 101-110.

Reply:
added sanity check